File: 7-luhn-algorithm/Haskell/7-luhn-algorithm.hs
module Main where

import Data.Time (getCurrentTime, diffUTCTime)

validate :: Integer -> Bool
validate = (== 0) . (`mod` 10) . sum . zipWith ($) (cycle [id, sum . digits . (*2)]) . reverse . digits
  where digits = map (read . return) . show

main :: IO ()
main = do
  start <- getCurrentTime
  
  print $ validate 2121 

  end <- getCurrentTime
  print (diffUTCTime end start)

File: 7-luhn-algorithm/JS/7-luhn-algorithm.js

function validate(n) {
  let sum = 0;

  while (n > 0) {
    let a = n % 10;
    n = Math.floor(n / 10);

    let b = (n % 10) * 2;
    n = Math.floor(n / 10);

    if (b > 9) b -= 9;
    sum += a + b;
  }

  return sum % 10 === 0;
}

function main() {
  console.time('Execution Time');
  console.log(validate(1354143611634));
  console.timeEnd('Execution Time');
}

main();

